trace/obfuscate.go

```diff
@@ -1,40 +1,204 @@
+}
+
+type redactSensitiveInfoVisitor struct {
+	vis.DefaultSpecVisitorImpl
+	obfuscationOptions *Obfuscator
+}
+
+var _ vis.DefaultSpecVisitor = (*redactSensitiveInfoVisitor)(nil)
+
+// EnterData processes the given data and obfuscates sensitive information based on the provided obfuscation options.
+// It does 3 checks to determine if the data contains sensitive information:
+// 1. It checks if spec is an HTTP Authroization or Cookie data and obfuscates the value.
+// 2. It cheks if spec is an HTTP Header or Query Param data and obfuscates the value if the key is in the list of sensitive keys.
+// 3. It applies regex patterns to all obfuscate sensitive primitive string values.
+func (s *redactSensitiveInfoVisitor) EnterData(self interface{}, ctx vis.SpecVisitorContext, d *pb.Data) Cont {
+	if httpMeta := d.GetMeta().GetHttp(); httpMeta != nil {
+		var key string
+		switch httpMeta.Location.(type) {
+		case *pb.HTTPMeta_Auth:
+			return ObfuscatePrimitiveWithRedactedString(d)
+		case *pb.HTTPMeta_Cookie:
+			return ObfuscatePrimitiveWithRedactedString(d)
+		case *pb.HTTPMeta_Header:
+			header := httpMeta.GetHeader()
+			key = header.Key
+		case *pb.HTTPMeta_Query:
+			queryParam := httpMeta.GetQuery()
+			key = queryParam.Key
+		}
+		// Check if the key is in the list of keys to obfuscate.
+		if s.obfuscationOptions.SensitiveDataKeys.Contains(strings.ToLower(key)) {
+			return ObfuscatePrimitiveWithRedactedString(d)
+		}
+	}
+
+	primitive := d.GetPrimitive()
+	if primitive == nil {
+		// Not a primitive, regex will be applied to primitive string values only.
+		return Continue
+	}
+
+	stringValue := primitive.GetStringValue()
+	if stringValue == nil {
+		// Not a string, regex will be applied to string values only.
+		return Continue
+	}
+
+	for pattern, _ := range s.obfuscationOptions.SensitiveDataValuePatterns {
+		if pattern.MatchString(stringValue.Value) {
+			return ObfuscatePrimitiveWithRedactedString(d)
+		}
+	}
+
+	return Continue
+}
+
+func (s *redactSensitiveInfoVisitor) EnterHTTPBody(self interface{}, ctx vis.SpecVisitorContext, b *pb.HTTPBody) Cont {
+	node, _ := ctx.GetInnermostNode(reflect.TypeOf((*pb.Data)(nil)))
+	data := node.(*pb.Data)
+	if data == nil {
+		return Continue
+	}
+
+	// Traverse the data node and redact sensitive information.
+	s.traverseAndRedactSensitiveInfo(ctx, data)
+
+	return Continue
+}
+
+func (s *redactSensitiveInfoVisitor) traverseAndRedactSensitiveInfo(ctx vis.SpecVisitorContext, data *pb.Data) {
+	switch val := data.Value.(type) {
+	case *pb.Data_Struct:
+		// Traverse the struct's fields and redact sensitive information based on key.
+		structData := val.Struct
+		for fieldName, fieldData := range structData.GetFields() {
+			// If field Data is primitive or primitiveList, then check it's key and redact if needed
+			if spec_util.IsPrimitive(fieldData) || spec_util.IsPrimitiveList(fieldData) {
+				if s.obfuscationOptions.SensitiveDataKeys.Contains(strings.ToLower(fieldName)) {
```

Comment:
Sorry, this isn't right yet.  You need to recursively visit non-primitive types in the fieldData.

Here's a test case for you.  Please write a unit test for this function.
```
{
  "api_key" : [
     "PMAK-1234", 
     {key_version: 3, key_value: "PMAK-5678"}
  ],
}
```

what we want is for every value inside api_key to be redacted:

```
{
  "api_key" : [
     "REDACTED", 
     {"key_version": "REDACTED", "key_value": "REDACTED"}
  ],
}
```

You should track during recursion whether there has been a sensitive field in the parent, and redact everything until we return back above that point.